### PR TITLE
[WIP] Add tool feedback/drop tool/magic edge tool/multi selection/diagram dirty state

### PR DESCRIPTION
**Pull request by hmarchadour**
### Type of this PR 

- [ ] Bug fix
- [x] Improvement
- [ ] Documentation
- [ ] Cleanup
- [ ] Test

### Issue(s)
 
 * #129
 * #131
 * #132 
 * #133 
 * #134 
 * #135

### What does this PR do?

**This PR depends on https://github.com/eclipse-sirius/sirius-components/pull/140**

A set of diagram enhancements :
* Add tool apply feedback
* Add a drop tool
* Add a magic edge tool
* Add multi selection
* Introduce a dirty boolean to detect a incoming refresh

### Screenshot/screencast of this PR

TODO

### Potential side effects

* Diagram refresh
* Delete tool
* Selection
* Invoke node/edge tool
* contextual palette

### How to test this PR?

- [ ] Frontend Unit tests
- [ ] Backend Unit tests
- [ ] Cypress : TODO?
- [ ] Manual Test : TODO?

### Checklist

- [x] I have read CONTRIBUTING carefully.
- [x] I have signed the [Contributor License Agreement](http://www.eclipse.org/legal/CLA.php).
- [x] All my commits are signed-off (`-s`) with my mail address of my Eclipse Account.
- [ ] I have covered my changes by unit tests or integration tests or manual tests.
- [ ] All tests pass.
- [ ] I have updated the documentation accordingly
- [ ] New React components are available in storybook

### CODE REVIEW

File: frontend/src/diagram/palette/ContextualPaletteContainer.tsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     Obeo - initial API and implementation
+ *******************************************************************************/
+import React, { CSSProperties } from 'react';
+import PropTypes from 'prop-types';
+import { EditLabelAction } from 'sprotty';
```
Comment (sbegaudeau) [inline frontend/src/diagram/palette/ContextualPaletteContainer.tsx:15]: Does this container need a dependency to Sprotty?

File: frontend/src/diagram/reducer.ts [hunk 2]
```diff
@@ -116,9 +132,10 @@ const switchRepresentationAction = (action) => {
   return {
     viewState: LOADING__STATE,
     displayedRepresentationId: action.representationId,
-    diagramServer: undefined,
+    modelSource: undefined,
     diagram: undefined,
     toolSections: [],
+    sprottyState: undefined,
```
Comment (sbegaudeau) [inline frontend/src/diagram/reducer.ts:146]: I don't know where I have mentioned this but do not name this "sprottyState".

**Review by sbegaudeau — CHANGES_REQUESTED** [review]
There is a lot of code in this review, I have not reviewed everything but here are a couple of comments to start the discussion.

File: backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/IDiagramDescriptionService.java [hunk 3]
```diff
@@ -27,4 +28,6 @@
 
     Optional<NodeDescription> findNodeDescriptionById(DiagramDescription diagramDescription, UUID nodeDescriptionId);
 
+    Optional<EdgeDescription> findEdgeDescriptionById(DiagramDescription diagramDescription, UUID edgeDescriptionId);
+
```
[the remark this answers is not in the document]
Comment (hmarchadour) [reply backend/sirius-web-collaborative-diagrams-api/src/main/java/org/eclipse/sirius/web/collaborative/diagrams/api/IDiagramDescriptionService.java:32]: Yes, I do. I should probably don't try to make distinct PR...

File: frontend/jsconfig.json [hunk 4]
```diff
@@ -1,7 +1,8 @@
 {
   "compilerOptions": {
     "baseUrl": "src",
-    "lib": ["es2020"]
+    "lib": ["es2020"],
+    "experimentalDecorators": true
```
[the remark this answers is not in the document]
Comment (hmarchadour) [reply frontend/jsconfig.json:5]: I've added this feature to deal with @inject decorator (already used in sprotty) but it's perhaps a bad idea.

File: frontend/src/diagram/Toolbar.tsx [hunk 5]
```diff
@@ -13,6 +13,8 @@
 import { IconButton } from 'core/button/Button';
 import { Select } from 'core/select/Select';
 import { FitToScreen, Share, ZoomIn, ZoomOut } from 'icons';
+import { FitToScreenAction } from 'sprotty';
+import { ZOOM_IN_ACTION, ZOOM_OUT_ACTION, ZOOM_TO_ACTION } from 'diagram/sprotty/Actions';
```
[the remark this answers is not in the document]
Comment (hmarchadour) [reply frontend/src/diagram/Toolbar.tsx:17]: Ok. The side effect of a "strict" appliance of this rule is a growing DiagramWebSocketContainer. The purpose of this change it to decrease the DiagramWebSocketContainer responsibility... I understand your point of view. I will review that.

File: frontend/src/diagram/palette/ContextualPaletteContainer.tsx (hunk 1 above)
Comment (hmarchadour) [reply frontend/src/diagram/palette/ContextualPaletteContainer.tsx:15]: I could reply the same thing https://github.com/eclipse-sirius/sirius-components/pull/143#discussion_r534007046

**Comment by hmarchadour** [comment]
I close this PR (I will open a big PR with all commits)

**Pull request closed without merging**